### ISSUE TITLE
Memory Leak with tf.Tensor.getitem in tf.data.Dataset

### ISSUE THREAD

**Issue by devstein** [tensorflow/tensorflow#30156]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): OSX and Linux
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: N/A
- TensorFlow installed from (source or binary): `pipenv install --pre tensorflow==2.0.0-beta1`
- TensorFlow version (use command below): `2.0.0-beta1`
- Python version: `3.6.8`

**Describe the current behavior**
As described in [tf.slice](https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/slice), I tried to use `tf.Tensor.getitem` to perform a slice in a more pythonic way, but ran into memory leaks. 

```python
def map_to_xy_dataset(csv_dataset, params):
    window_size = params[WINDOW_SIZE_KEY]
    window_shift = params[WINDOW_SHIFT_KEY]
    n_workers = tf.data.experimental.AUTOTUNE

    # MEMORY LEAK IN THIS FUNCTION
    def split_xy(window):
        # For X, select all but the last value and flatten
        range_x = tf.shape(window)[1] - 1
        
        # CAUSES MEMORY LEAK 
        x = tf.reshape(window[:, 0:range_x], [-1])


        # Select a single Y value
        # CAUSES MEMORY LEAK
        y = window[0, -1]

        return x, y

    # Turn the csv dataset row x col tensor
    row_dataset = csv_dataset.map(lambda *x: tf.reshape(x, [len(x)]))

    windowed_dataset = row_dataset.window(
        size=window_size, shift=window_shift,
        drop_remainder=True).flat_map(lambda x: x.batch(window_size))

    xy_dataset = windowed_dataset.map(split_xy, num_parallel_calls=n_workers)

    return xy_dataset
```

I have **not** tried to reproduce this behavior with  [tf.slice](https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/slice). I will update this issue if I do. 

**Describe the expected behavior**
No memory leaks.

**Code to reproduce the issue**
See above.

**Other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

Initial related comment: https://github.com/tensorflow/tensorflow/issues/19049#issuecomment-505585437

**Comment by achandraa** [comment]
Will it be possible to provide a full code snippet that can replicate the issue.It will indeed help us to move faster. Thanks!

**Comment by devstein** [comment]
@achandraa  Thanks for looking into this! Does this code snippet suffice? If not, let me know what other details you need.

```python
import tensorflow as tf


def map_to_xy_dataset(csv_dataset, params):
    window_size = 3
    window_shift = 10
    n_workers = tf.data.experimental.AUTOTUNE

    # MEMORY LEAK IN THIS FUNCTION
    def split_xy(window):
        # For X, select all but the last value and flatten
        range_x = tf.shape(window)[1] - 1
        
        # CAUSES MEMORY LEAK 
        x = tf.reshape(window[:, 0:range_x], [-1])


        # Select a single Y value
        # CAUSES MEMORY LEAK
        y = window[0, -1]

        return x, y

    # Turn the csv dataset row x col tensor
    row_dataset = csv_dataset.map(lambda *x: tf.reshape(x, [len(x)]))

    windowed_dataset = row_dataset.window(
        size=window_size, shift=window_shift,
        drop_remainder=True).flat_map(lambda x: x.batch(window_size))

    xy_dataset = windowed_dataset.map(split_xy, num_parallel_calls=n_workers)

    return xy_dataset

def generate_dataset():
    filenames_dataset = tf.data.Dataset.list_files(data_glob)

    dataset = filenames_dataset.flat_map(map_file_to_xy_dataset)

    return dataset


dataset = generate_dataset()

model = ...

model.fit(dataset, ...)
```

**Comment by gadagashwini** [comment]
@devstein I have tried to reproduce the issue with the above code but Looks like some of the entities like data_glob, map_file_to_xy_dataset are not defined. Can you help us to reproduce the issue. Thanks!

**Comment by devstein** [comment]
@gadagashwini Here is a more detailed example: 

```python
import tensorflow as tf

def map_file_to_xy_dataset(filename):
    window_size = 3
    window_shift = 10
    n_workers = tf.data.experimental.AUTOTUNE

    # Generate a dataset from the filename
    csv_dataset = tf.data.experimental.CsvDataset("Iris.csv",["float32","float32","float32","float32","float32"])

    # Cache at unique file location (i.e filename)
    cached_dataset = csv_dataset.cache("Iris.csv")

    # MEMORY LEAK IN THIS FUNCTION
    def split_xy(window):
        # For X, select all but the last value and flatten
        range_x = tf.shape(window)[1] - 1
        
        # CAUSES MEMORY LEAK 
        x = tf.reshape(window[:, 0:range_x], [-1])


        # Select a single Y value
        # CAUSES MEMORY LEAK
        y = window[0, -1]

        return x, y

    # Turn the csv dataset row x col tensor
    row_dataset = cached_dataset.map(lambda *x: tf.reshape(x, [len(x)]))

    windowed_dataset = row_dataset.window(
        size=window_size, shift=window_shift,
        drop_remainder=True).flat_map(lambda x: x.batch(window_size))

    xy_dataset = windowed_dataset.map(split_xy, num_parallel_calls=n_workers)

    return xy_dataset


def generate_dataset():
    filenames_dataset = tf.data.Dataset.list_files("Iris.csv")

    dataset = filenames_dataset.flat_map(map_file_to_xy_dataset)

    return dataset


```

**Comment by devstein** [comment]
@gadagashwini After more experimenting, I don't think `tf.Tensor.getitem` is the root of the issue. This [issue](https://github.com/tensorflow/tensorflow/issues/29697) describes the same behavior but is only caching. 

Also, this issue is only persistent during the first epoch. Does `tf.data.Dataset.cache` cache to memory before it writes to disk?

**Comment by gadagashwini** [comment]
@devstein I tried reproducing the issue with the above provided code but it didn't show any error. Please provide us complete code which replicates the reported issue. Thanks!

**Comment by devstein** [comment]
@gadagashwini It causes memory usage climbs, see my last comment. There is no error produced.

**Comment by jsimsa** [comment]
@devstein the effect of `cycle_length` with `interleave` is that a multiple of `cycle_length` instances of the input pipeline returned by `map_file_to_xy_dataset` will be evaluated in parallel. If each instances maintains internal buffers, the memory usage will be multiplied by the `cycle_length` factor.

In your case, it looks like your input pipeline is using `window` and `map` with `num_parallel_calls`. Both of these transformations will maintain an internal buffer.

In other words, the behavior you are experiencing is not a memory leak (at least not based on the evidence you have provided), but high memory usage.

**Comment by devstein** [comment]
@jsimsa That makes sense, but why is this only prevalent when using `.cache(...)`? Memory usage is significantly lower when using the exact same parameters but not caching each `csv_dataset` to disk?

**Comment by jsimsa** [comment]
IIRC, the implementation of cache would uses an internal buffer for batching writes of a tensor. So that would explain why is the problem exacerbated by using `cache` in your input pipeline.
